{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e77d9531-bb96-4a61-a97e-a3aaa11b826d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "modelseedpy 0.4.2\n",
      "cobrakbase 0.4.0\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "from pathlib import Path\n",
    "sys.path.append(\"/kb/module/berdl\")\n",
    "from berdl.genome_paths import GenomePaths\n",
    "from berdl.prep_genome_set import BERDLPreGenome\n",
    "from cobrakbase import KBaseAPI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a8e94901-b3a2-4f2f-af90-cdd5f1ba4e09",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "paths = GenomePaths(root=Path(\"/test/syncom\").resolve())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9067e829-fe07-4f5a-a89e-3ae3e93256b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "kbase = KBaseAPI('JLQJGZGRPPXV2VPTJJM6HFWM4NRK46QW')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1f41a1ed-ea4e-409d-ba8e-7e13dd0c60fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_R12 = kbase.get_from_ws('158561/3/1')\n",
    "genome_3H11 = kbase.get_from_ws('158561/6/1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "73560371-3bb3-4d59-a7a9-45c8b2940891",
   "metadata": {},
   "outputs": [],
   "source": [
    "berdl_prep_genomes = BERDLPreGenome(kbase, paths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "488e1576-0ee7-4aee-9b99-8dc5967033ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "user_to_clade, ani_clade, df_ani_fitness, df_ani_phenotype = berdl_prep_genomes.run([genome_R12, genome_3H11])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "704504a7-54dd-4dfc-b9b8-2990a6a3ff83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Rhodanobacter-sp.FW510-R12': {'RS_GCF_021545845.1': [100.0, 100.0, 100.0],\n",
       "  'RS_GCF_000264375.1': [94.18, 73.35, 70.98],\n",
       "  'RS_GCF_000230695.2': [90.73, 57.07, 57.36],\n",
       "  'RS_GCF_003069545.1': [89.92, 61.33, 51.62],\n",
       "  'RS_GCF_001617365.1': [88.84, 50.09, 44.29],\n",
       "  'RS_GCF_018599185.1': [88.27, 50.1, 45.62],\n",
       "  'GB_GCA_001898415.1': [85.46, 20.18, 19.57],\n",
       "  'RS_GCF_016889805.1': [85.31, 23.64, 22.06]},\n",
       " 'Acidovorax-sp.GW101-3H11': {'RS_GCF_001633105.1': [99.99, 99.82, 99.21],\n",
       "  'RS_GCF_002245625.1': [93.76, 81.44, 76.42],\n",
       "  'RS_GCF_000302535.1': [88.27, 46.25, 43.89],\n",
       "  'RS_GCF_007993815.1': [88.08, 46.43, 42.58],\n",
       "  'RS_GCF_023913775.1': [87.81, 32.99, 38.56],\n",
       "  'RS_GCF_003058055.1': [87.8, 42.48, 39.66],\n",
       "  'RS_GCF_019218755.1': [86.44, 30.56, 32.14],\n",
       "  'RS_GCF_000204195.1': [86.33, 32.25, 32.56],\n",
       "  'RS_GCF_002176815.1': [85.35, 15.89, 12.82],\n",
       "  'RS_GCF_003669015.1': [85.28, 15.52, 14.3],\n",
       "  'RS_GCF_001424425.1': [85.23, 22.42, 24.04],\n",
       "  'GB_GCA_018828125.1': [84.52, 18.75, 17.72]}}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ani_clade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4e4df507-7bd9-4a92-98c7-12e83fffeb12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Rhodanobacter-sp.FW510-R12': 'RS_GCF_021545845.1',\n",
       " 'Acidovorax-sp.GW101-3H11': 'RS_GCF_001633105.1'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "user_to_clade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fa99921f-779e-4531-8878-8ec500a7049c",
   "metadata": {},
   "outputs": [],
   "source": [
    "clade_to_user_genomes = {}\n",
    "for u, c in user_to_clade.items():\n",
    "    if c not in clade_to_user_genomes:\n",
    "        clade_to_user_genomes[c] = set()\n",
    "    clade_to_user_genomes[c].add(u)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fed7a59a-f08e-4404-b823-083c6ab83f1c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c78b084-24d9-4db6-9d26-0b1c25c881c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "pg_query = QueryPangenomes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "df2cba88-298c-4e82-874d-b1eb7a71eff9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from berdl.berdl_api import BERDLAPI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "00379b14-459c-44e4-82ec-77e30b15a167",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 24.4 ms, sys: 2.24 ms, total: 26.7 ms\n",
      "Wall time: 55.1 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "berdl_token = \"5AROMD2TY337O2E2AGOL52HL5T7EPKT6\"\n",
    "berdl_api = BERDLAPI(berdl_token)\n",
    "berdl_api.timeout = 12000\n",
    "genome_ke_member_id = 'GB_GCA_019459165.1'\n",
    "sql = f\"SELECT * FROM kbase_genomes.name WHERE name = '{genome_ke_member_id}' LIMIT 1\"\n",
    "response = berdl_api.sql(sql, offset=0, page_limit=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "44e275d5-ec1b-410d-938f-1b75198ccdc3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "b'{\"error\":40800,\"error_type\":\"request_timeout\",\"message\":\"Request timed out after 55.0 seconds. The operation took too long to complete. Consider using pagination, reducing query scope, or breaking up large operations into smaller requests.\"}'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response.content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "272515a3-d134-426d-b355-bd538bcba50a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b6779a87-5203-4ffd-94af-6fae5aadd19b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "contig.parquet\n",
      "contig_collection.parquet\n",
      "contig_collection_x_encoded_feature.parquet\n",
      "contig_collection_x_feature.parquet\n",
      "contig_collection_x_protein.parquet\n",
      "contig_x_contig_collection.parquet\n",
      "contig_x_encoded_feature.parquet\n",
      "contig_x_feature.parquet\n",
      "contig_x_protein.parquet\n",
      "encoded_feature.parquet\n",
      "encoded_feature_x_feature.parquet\n",
      "encoded_feature_x_protein.parquet\n",
      "feature.parquet\n",
      "feature_name.parquet\n",
      "feature_prodigal_attributes.parquet\n",
      "feature_x_protein.parquet\n",
      "mc_upload_tenant_berdl.sh\n",
      "mc_upload_user_berdl.sh\n",
      "name.parquet\n",
      "protein.parquet\n"
     ]
    }
   ],
   "source": [
    "!ls /data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "29aa42f0-8e96-46dd-961a-f285a9224ffe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl\n",
    "from pathlib import Path\n",
    "from berdl.query.query_genome import QueryGenomeABC\n",
    "\n",
    "\n",
    "class QueryGenomeLocal(QueryGenomeABC):\n",
    "\n",
    "    def __init__(self):\n",
    "        self.root_reference = Path('/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes')\n",
    "        self.ldf_name = pl.scan_parquet(f'{self.root_reference}/block_*/name.parquet')\n",
    "        pass\n",
    "\n",
    "    def get_genome_features(self, genome_id: str):\n",
    "        \"\"\"Return all gene clusters for a given clade.\"\"\"\n",
    "        raise NotImplementedError\n",
    "\n",
    "    def get_genome_contigs(self, genome_id: str):\n",
    "        \"\"\"Return all gene clusters for a given clade.\"\"\"\n",
    "        raise NotImplementedError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "10256923-bc7c-4707-892c-4e07fed094cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "qg = QueryGenomeLocal()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "c93486c6-da91-471b-9f93-9c95d935211b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Schema([('entity_id', String),\n",
       "        ('name', String),\n",
       "        ('description', String),\n",
       "        ('source', String)])"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qg.ldf_name.collect_schema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "de482924-dac6-4280-a2c3-ade779f45f2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 927 ms, sys: 296 ms, total: 1.22 s\n",
      "Wall time: 51.3 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "df = qg.ldf_name.filter(pl.col(\"name\") == genome_ke_member_id).collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "1b4977e0-9488-444a-b29b-1e143ea5e943",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.height"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "79d4a53f-0e30-4f57-82bd-136f684d0f23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'entity_id': 'CDM:ccol-9034b761-4f23-4bc4-ad6b-54e7ec40366a',\n",
       " 'name': 'GB_GCA_019459165.1',\n",
       " 'description': 'gtdb_r214',\n",
       " 'source': 'ke-pangenomes'}"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.row(0, named=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05b8e5c9-7f6d-49c9-b9bd-f29da00b2965",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (berdl_genomes)",
   "language": "python",
   "name": "berdl_genomes"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
