{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "40169956-d58a-4a81-9bb5-c45337f0df57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "modelseedpy 0.4.2\n",
      "cobrakbase 0.4.0\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "from pathlib import Path\n",
    "sys.path.append(\"/kb/module/berdl\")\n",
    "from berdl.genome_paths import GenomePaths\n",
    "from berdl.prep_genome_set import BERDLPreGenome\n",
    "from cobrakbase import KBaseAPI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "7c812a69-3488-4a14-afa4-cbb9c0aab85e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "336"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import polars as pl\n",
    "pl.thread_pool_size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "94a9b47c-e83a-4dd3-bc79-91d537c0a571",
   "metadata": {},
   "outputs": [],
   "source": [
    "from berdl.query.query_pangenome_local import QueryPangenomeLocal\n",
    "from berdl.query.query_genome_local import QueryGenomeLocal\n",
    "from berdl.pangenome.pangenome import BERDLPangenome\n",
    "from berdl.pangenome.paths_pangenome import PathsPangenome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4d8bb305-7b1c-4249-aca1-59c9ed039dbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "scratch = '/kb/module/work/tmp'\n",
    "root_pangenome = Path(scratch) / 'pangenome' / 'GB_GCA_021307345.1'\n",
    "paths = PathsPangenome(root=root_pangenome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1212d8cc-5c08-490c-a603-0e34c3f38a1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_pg = QueryPangenomeLocal('/data/reference_data/berdl_db/ke-pangenomes')\n",
    "query_g = QueryGenomeLocal()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c821cce-bb63-4567-b212-88e5c26d0161",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "add6a9fc-963b-43e5-a16a-dce797e5d3fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from modelseedpy.core.msgenome import MSFeature, MSGenome\n",
    "from berdl.pangenome.paths_pangenome import PathsPangenome\n",
    "\n",
    "\n",
    "class BERDLPangenome:\n",
    "\n",
    "    def __init__(self, query_pg, query_g, paths: PathsPangenome):\n",
    "        self.pg = query_pg  # pan-genome query api\n",
    "        self.query_g = query_g\n",
    "        self.paths = paths.ensure()\n",
    "\n",
    "    def run(self, genome_id):\n",
    "        clade_id = self.pg.get_member_representative(genome_id)\n",
    "        clade_members = self.pg.get_clade_members(clade_id)\n",
    "        clade_gene_clusters = self.pg.get_clade_gene_clusters(clade_id)\n",
    "        clade_cluster_ids = set(clade_gene_clusters['gene_cluster_id'])\n",
    "        df_gene_genecluster = self.pg.get_clusters_members(clade_cluster_ids)\n",
    "        d_gene_to_cluster = {o[0]: o[1] for o in df_gene_genecluster.iter_rows()}\n",
    "        d_cluster_to_genes = {}\n",
    "        for k, v in d_gene_to_cluster.items():\n",
    "            if v not in d_cluster_to_genes:\n",
    "                d_cluster_to_genes[v] = set()\n",
    "            d_cluster_to_genes[v].add(k)\n",
    "\n",
    "        # get clade member faa and fna\n",
    "        members = {}\n",
    "        for row in clade_members.rows(named=True):\n",
    "            member_id = row['genome_id']\n",
    "            members[member_id] = self.query_g.get_genome_features(member_id)\n",
    "\n",
    "        # build master protein user_genome + pangenome\n",
    "\n",
    "        u_proteins = {}\n",
    "        for k, g in members.items():\n",
    "            print(k, len(g.features))\n",
    "            for feature in g.features:\n",
    "                _parts = feature.description.split(' ')\n",
    "                h = _parts[4]\n",
    "                u_proteins[h] = MSFeature(h, feature.seq)\n",
    "\n",
    "        genome_master_faa = MSGenome()\n",
    "        genome_master_faa.add_features(list(u_proteins.values()))\n",
    "        genome_master_faa.to_fasta(str(self.paths.out_master_faa_pangenome_members))\n",
    "\n",
    "        #  collect user genome and add to u_proteins\n",
    "        \n",
    "\n",
    "        #  collect phenotype and fitness\n",
    "        genome_master_faa_fitness = MSGenome.from_fasta(str(self.paths.ref_master_faa_protein_fitness))\n",
    "        for feature in genome_master_faa_fitness:\n",
    "            if feature.id not in u_proteins:\n",
    "                u_proteins[feature.id] = MSFeature(feature.id, feature.seq)\n",
    "        genome_master_faa_phenotype = MSGenome.from_fasta(str(self.paths.ref_master_faa_protein_phenotype))\n",
    "        for feature in genome_master_faa_phenotype:\n",
    "            if feature.id not in u_proteins:\n",
    "                u_proteins[feature.id] = MSFeature(feature.id, feature.seq)\n",
    "\n",
    "        # rebuild master faa genome with proteins from\n",
    "        #  user / pangenome / fitness / phenotypes\n",
    "        genome_master_faa = MSGenome()\n",
    "        genome_master_faa.add_features(list(u_proteins.values()))\n",
    "        genome_master_faa.to_fasta(str(self.paths.out_master_faa))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "6b0b665f-8d32-4223-813f-9c3d8f2eb82c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_0/feature_name.parquet\n",
      "/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_0/name.parquet\n",
      "/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_1/name.parquet\n",
      "/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_1/name_feature.parquet\n",
      "/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_2/name.parquet\n",
      "/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_2/name_feature.parquet\n",
      "/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_3/name.parquet\n",
      "/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_3/name_feature.parquet\n",
      "/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_4/name.parquet\n",
      "/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_4/name_feature.parquet\n",
      "/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_5/name.parquet\n",
      "/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_5/name_feature.parquet\n"
     ]
    }
   ],
   "source": [
    "import polars as pl\n",
    "from pathlib import Path\n",
    "\n",
    "paths = sorted(Path('/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes').glob(\"block_*/*name*.parquet\"))\n",
    "\n",
    "for p in paths:\n",
    "    n = pl.scan_parquet(str(p)).select(pl.len()).collect().item()\n",
    "    print(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6a037ff5-1db7-48b4-9f04-c1a45e8c06be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parquet SCAN [/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_0/feature_name.parquet, ... 11 other sources]\n",
      "PROJECT */4 COLUMNS\n",
      "ESTIMATED ROWS: 2281548108\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_20155/1651664276.py:1: DeprecationWarning: the `streaming` parameter was deprecated in 1.25.0; use `engine` instead.\n",
      "  print(query_g.ldf_name.explain(streaming=True))\n"
     ]
    }
   ],
   "source": [
    "print(query_g.ldf_name.explain(streaming=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "f9ff9d72-72a5-4c81-9dd4-96a47d5f5565",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Parquet SCAN [/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_0/name.parquet, ... 13 other sources]\\nPROJECT */4 COLUMNS\\nESTIMATED ROWS: 61191284'"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ldf_name = pl.scan_parquet(\n",
    "    [\n",
    "        '/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_0/name.parquet',\n",
    "        '/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_1/name.parquet',\n",
    "        '/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_2/name.parquet',\n",
    "        '/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_3/name.parquet',\n",
    "        '/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_4/name.parquet',\n",
    "        '/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_5/name.parquet',\n",
    "        '/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_0/feature_name.parquet',\n",
    "        '/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_1/name_feature.parquet',\n",
    "        '/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_2/name_feature.parquet',\n",
    "        '/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_3/name_feature.parquet',\n",
    "        '/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_4/name_feature.parquet',\n",
    "        '/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/block_5/name_feature.parquet',\n",
    "        '/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/name.parquet',\n",
    "        '/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes/name_feature.parquet',\n",
    "    ]\n",
    ")\n",
    "ldf_name.explain()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "a9b71295-4725-4c39-9b17-815332a28a5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (2, 1)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>entity_id</th></tr><tr><td>str</td></tr></thead><tbody><tr><td>&quot;CDM:ccol-b754b396-c86c-41e4-ad…</td></tr><tr><td>&quot;CDM:ccol-b754b396-c86c-41e4-ad…</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (2, 1)\n",
       "┌─────────────────────────────────┐\n",
       "│ entity_id                       │\n",
       "│ ---                             │\n",
       "│ str                             │\n",
       "╞═════════════════════════════════╡\n",
       "│ CDM:ccol-b754b396-c86c-41e4-ad… │\n",
       "│ CDM:ccol-b754b396-c86c-41e4-ad… │\n",
       "└─────────────────────────────────┘"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = ldf_name.filter(pl.col(\"name\") == 'GB_GCA_019839465.1').select(\"entity_id\").collect()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "4d664849-cd39-4c8a-b801-3423bc6a3af5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CDM:ccol-b754b396-c86c-41e4-ada9-c416a4761802'}"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "486d4ff6-3c16-4072-81a5-fc89b9726c8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'entity_id': 'CDM:ccol-9034b761-4f23-4bc4-ad6b-54e7ec40366a',\n",
       " 'name': 'GB_GCA_019459165.1',\n",
       " 'description': 'gtdb_r214',\n",
       " 'source': 'ke-pangenomes'}"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_g.get_cdm_genome_id_by_name('GB_GCA_019459165.1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "caec933a-1643-4411-b486-2666c0afe9b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open(paths.genome_prep_clade_data, 'r') as fh:\n",
    "    clade_assignment = json.load(fh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e7107acb-3359-447f-87f5-60f466ecb52c",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_clade_member_id = 'GB_GCA_021307345.1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "15a7a491-8e7d-4ce7-8a38-c264d9c58a43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Escherichia_coli_K-12_MG1655'}"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{k for k, v in clade_assignment.items() if v == selected_clade_member_id}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a5657e64-9c1f-49f3-91e6-b4311f40e4e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "berld_pangenome = BERDLPangenome(query_pg, query_g, paths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "efd4ede6-f59b-45a4-8a05-45bba1139cb9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GB_GCA_019459165.1 5983\n",
      "RS_GCF_001633105.1 4955\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "'MSGenome' object is not iterable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m member_id \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mRS_GCF_001633105.1\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mberld_pangenome\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmember_id\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[9], line 50\u001b[0m, in \u001b[0;36mBERDLPangenome.run\u001b[0;34m(self, genome_id)\u001b[0m\n\u001b[1;32m     45\u001b[0m \u001b[38;5;66;03m#  collect user genome and add to u_proteins\u001b[39;00m\n\u001b[1;32m     46\u001b[0m \n\u001b[1;32m     47\u001b[0m \n\u001b[1;32m     48\u001b[0m \u001b[38;5;66;03m#  collect phenotype and fitness\u001b[39;00m\n\u001b[1;32m     49\u001b[0m genome_master_faa_fitness \u001b[38;5;241m=\u001b[39m MSGenome\u001b[38;5;241m.\u001b[39mfrom_fasta(\u001b[38;5;28mstr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpaths\u001b[38;5;241m.\u001b[39mref_master_faa_protein_fitness))\n\u001b[0;32m---> 50\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m feature \u001b[38;5;129;01min\u001b[39;00m genome_master_faa_fitness:\n\u001b[1;32m     51\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m feature\u001b[38;5;241m.\u001b[39mid \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m u_proteins:\n\u001b[1;32m     52\u001b[0m         u_proteins[feature\u001b[38;5;241m.\u001b[39mid] \u001b[38;5;241m=\u001b[39m MSFeature(feature\u001b[38;5;241m.\u001b[39mid, feature\u001b[38;5;241m.\u001b[39mseq)\n",
      "\u001b[0;31mTypeError\u001b[0m: 'MSGenome' object is not iterable"
     ]
    }
   ],
   "source": [
    "member_id = 'GB_GCA_021307345.1'\n",
    "berld_pangenome.run(member_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c3c89fd0-84cf-4719-8751-2e9bd49c8424",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'s__Acidovorax_sp001633105--RS_GCF_001633105.1'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clade_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b8553647-97c9-4209-8230-c5da9d892d1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "members = {}\n",
    "for row in clade_members.rows(named=True):\n",
    "    member_id = row['genome_id']\n",
    "    members[member_id] = berld_pangenome.query_g.get_genome_features(member_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "cec26ef1-4482-4986-9a2d-726e33057230",
   "metadata": {},
   "outputs": [],
   "source": [
    "from modelseedpy.core.msgenome import MSFeature, MSGenome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f587cb5b-5e5d-4137-9a5a-c522f7cba51d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GB_GCA_019459165.1 5983\n",
      "RS_GCF_001633105.1 4955\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "9997"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "u_proteins = {}\n",
    "for k, g in members.items():\n",
    "    print(k, len(g.features))\n",
    "    for feature in g.features:\n",
    "        _parts = feature.description.split(' ')\n",
    "        h = _parts[4]\n",
    "        u_proteins[h] = MSFeature(h, feature.seq)\n",
    "len(u_proteins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "12872ca3-7c2c-4019-90bb-83f3b563d093",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2861cd6923265c67fd04b835d48aa89c383f5cae645e53c14ee338e3cc4a56f1'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bc0a51a6-f154-4dbb-adaa-7977f5c648da",
   "metadata": {},
   "outputs": [],
   "source": [
    "from berdl.hash_seq import ProteinSequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "47df8485-4271-4a8d-8508-e6e27db0ad0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "clade_member_ids = {t[0] for t in clade_members.rows()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e1f58198-c3f9-4d38-a976-8aa0754298ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GB_GCA_019459165.1', 'RS_GCF_001633105.1'}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clade_member_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aea6e1a-81a1-4e57-9c71-14c48013ef55",
   "metadata": {},
   "outputs": [],
   "source": [
    "pl.scan_parquet(f'{self.root_reference}/block_*/name.parquet')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "981e656f-a626-40da-b75f-7a2ab05b3ead",
   "metadata": {},
   "outputs": [],
   "source": [
    "CDM:ccol-9034b761-4f23-4bc4-ad6b-54e7ec40366a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "b083b31a-e46b-4ccb-b519-0ce645bda1ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl\n",
    "from pathlib import Path\n",
    "from berdl.query.query_genome import QueryGenomeABC\n",
    "\n",
    "\n",
    "class QueryGenomeLocal(QueryGenomeABC):\n",
    "\n",
    "    def __init__(self):\n",
    "        self.root_reference = Path('/data/reference_data/berdl_db/cdm_genomes/ke-pangenomes')\n",
    "        self.ldf_name = pl.scan_parquet(f'{self.root_reference}/block_*/*name*.parquet')\n",
    "        #self.ldf_name_feature = pl.scan_parquet(f'{self.root_reference}/block_*/name_feature.parquet')\n",
    "        self.ldf_ccol_x_protein = pl.scan_parquet(f'{self.root_reference}/block_*/contig_collection_x_protein.parquet')\n",
    "        self.ldf_ccol_x_feature = pl.scan_parquet(f'{self.root_reference}/block_*/contig_collection_x_feature.parquet')\n",
    "        self.ldf_contig_x_feature = pl.scan_parquet(f'{self.root_reference}/block_*/contig_x_feature.parquet')\n",
    "        self.ldf_protein = pl.scan_parquet(f'{self.root_reference}/block_*/protein.parquet')\n",
    "        self.ldf_feature = pl.scan_parquet(f'{self.root_reference}/block_*/feature.parquet')\n",
    "        self.ldf_feature_x_protein = pl.scan_parquet(f'{self.root_reference}/block_*/feature_x_protein.parquet')\n",
    "\n",
    "    def get_cdm_genome_id_by_name(self, name: str):\n",
    "        df = self.ldf_name.filter(pl.col(\"name\") == name).collect()\n",
    "        if df.height == 1:\n",
    "            return df.row(0, named=True)\n",
    "        return None\n",
    "\n",
    "    def get_genome_features(self, genome_id: str):\n",
    "        res = self.get_cdm_genome_id_by_name(genome_id)\n",
    "        if res:\n",
    "            ccol_id = res['entity_id']\n",
    "            df = self.ldf_ccol_x_feature.filter(pl.col(\"contig_collection_id\") == ccol_id).select(\n",
    "                pl.col(\"feature_id\")).collect()\n",
    "            cdm_feat_ids = {o[0] for o in df.rows()}\n",
    "\n",
    "            df_features = self.ldf_feature.filter(pl.col(\"feature_id\").is_in(cdm_feat_ids)).select(\n",
    "                [\"feature_id\", \"start\", \"end\", \"type\"]).collect()\n",
    "\n",
    "            df_feature_to_contig = self.ldf_contig_x_feature.filter(\n",
    "                pl.col(\"feature_id\").is_in(cdm_feat_ids)).collect()\n",
    "            d_feature_to_contig = {row['feature_id']: row['contig_id'] for row in df_feature_to_contig.rows(named=True)}\n",
    "\n",
    "            df_feature_to_protein = self.ldf_feature_x_protein.filter(\n",
    "                pl.col(\"feature_id\").is_in(cdm_feat_ids)).collect()\n",
    "            d_feature_to_protein = {row['feature_id']: row['protein_id'] for row in\n",
    "                                    df_feature_to_protein.rows(named=True)}\n",
    "\n",
    "            cdm_prot_ids = set(d_feature_to_protein.values())\n",
    "            df_proteins = self.ldf_protein.filter(pl.col(\"protein_id\").is_in(cdm_prot_ids)).collect()\n",
    "            cdm_proteins = {}\n",
    "            for row in df_proteins.rows(named=True):\n",
    "                cdm_proteins[row['protein_id']] = row\n",
    "\n",
    "            cntg_ids = set(d_feature_to_contig.values())\n",
    "            df_contig_names = self.ldf_name.filter(pl.col(\"entity_id\").is_in(cntg_ids)).select(\n",
    "                [\"entity_id\", \"name\"]).collect()\n",
    "            d_cntg_id_to_name = {row['entity_id']: row['name'] for row in df_contig_names.rows(named=True)}\n",
    "\n",
    "            df_names = self.ldf_name.filter(pl.col(\"entity_id\").is_in(cdm_feat_ids)).select(\n",
    "                [\"entity_id\", \"name\"]).collect()\n",
    "\n",
    "            d_names = {row['entity_id']: row['name'] for row in df_names.rows(named=True)}\n",
    "\n",
    "            from modelseedpy.core.msgenome import MSFeature, MSGenome\n",
    "            features = {}\n",
    "            for row in df_features.rows(named=True):\n",
    "                start = row['start']\n",
    "                end = row['end']\n",
    "                feature_type = row['type']\n",
    "                cdm_feature_id = row['feature_id']\n",
    "                cdm_protein_id = d_feature_to_protein[cdm_feature_id]\n",
    "                cdm_protein = cdm_proteins[cdm_protein_id]\n",
    "                protein_hash = cdm_protein['hash']\n",
    "                sequence = cdm_protein['sequence']\n",
    "                cdm_contig_id = d_feature_to_contig[cdm_feature_id]\n",
    "                contig_name = d_cntg_id_to_name[cdm_contig_id]\n",
    "                feature_to_kpg_id = d_names.get(cdm_feature_id, cdm_feature_id)\n",
    "                desc = f'{contig_name} {start} {end} {feature_type} {protein_hash} {cdm_contig_id} {cdm_feature_id} {cdm_protein_id}'\n",
    "                genome_feature = MSFeature(feature_to_kpg_id, sequence, description=desc)\n",
    "                features[genome_feature.id] = genome_feature\n",
    "\n",
    "            genome = MSGenome()\n",
    "            genome.add_features(list(features.values()))\n",
    "            return genome\n",
    "\n",
    "        else:\n",
    "            return None\n",
    "\n",
    "    def get_genome_contigs(self, genome_id: str):\n",
    "        \"\"\"Return all gene clusters for a given clade.\"\"\"\n",
    "        raise NotImplementedError\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "8dd93311-647e-480f-b57d-a1b6cf1bbab0",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_genome = QueryGenomeLocal()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "69b7d43f-479e-4ade-8b40-f7583ab961b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_member = query_genome.get_genome_features('GB_GCA_019459165.1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "855753e5-da00-43a0-a14c-976d00054d94",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "JAIBEQ010000430.1_1 JAIBEQ010000430.1 1 2145 CDS 1476cb0d2f1f6946951c4bf697ae4aaa7a09cc669ac3350465a5baa45eacc778 CDM:cntg-0652c309-15ab-47fc-9c7a-fba4528b5d96 CDM:feat-7789e6da-eab5-46c7-9107-7add5ae62185 CDM:prot-de23b6c3-c0e3-4df9-95e0-0c6e6bd10e6e\n"
     ]
    }
   ],
   "source": [
    "for f in genome_member.features:\n",
    "    print(f.id, f.description)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "6ab606f3-6fb4-4068-a7fe-3d2ab068ceb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = query_genome.ldf_ccol_x_feature.filter(pl.col(\"contig_collection_id\") == ccol_id).select(pl.col(\"feature_id\")).collect()\n",
    "cdm_feat_ids = {o[0] for o in df.rows()}\n",
    "df_features = query_genome.ldf_feature.filter(pl.col(\"feature_id\").is_in(cdm_feat_ids)).select([\"feature_id\", \"start\", \"end\", \"type\"]).collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "5df773e8-ed96-471a-b5fe-ad035f02c44a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from modelseedpy.core.msgenome import MSFeature, MSFeature\n",
    "features = {}\n",
    "for row in df_features.rows(named=True):\n",
    "    start = row['start']\n",
    "    end = row['end']\n",
    "    feature_type = row['type']\n",
    "    cdm_feature_id = row['feature_id']\n",
    "    cdm_protein_id = d_feature_to_protein[cdm_feature_id]\n",
    "    cdm_protein = cdm_proteins[cdm_protein_id]\n",
    "    contig_id = d_feature_to_contig[cdm_feature_id]\n",
    "    protein_hash = cdm_protein['hash']\n",
    "    sequence = cdm_protein['sequence']\n",
    "    feature_to_kpg_id = d_names.get(cdm_feature_id, cdm_feature_id)\n",
    "    desc = f'{start} {end} {feature_type}'\n",
    "    MSFeature(feature_to_kpg_id, sequence, description=)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "4034b391-61c1-42fb-aa1e-e5f8669b3d19",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "9def42cc-e7a3-414d-ac7d-0edd5573b469",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (1_064, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>entity_id</th><th>name</th></tr><tr><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;CDM:cntg-0652c309-15ab-47fc-9c…</td><td>&quot;JAIBEQ010000430.1&quot;</td></tr><tr><td>&quot;CDM:cntg-77920f40-360f-4b30-b7…</td><td>&quot;JAIBEQ010000456.1&quot;</td></tr><tr><td>&quot;CDM:cntg-35e9d20a-1730-471b-8e…</td><td>&quot;JAIBEQ010000495.1&quot;</td></tr><tr><td>&quot;CDM:cntg-a142609d-adf6-4587-8c…</td><td>&quot;JAIBEQ010000514.1&quot;</td></tr><tr><td>&quot;CDM:cntg-3cecc5bf-5c5b-440c-b6…</td><td>&quot;JAIBEQ010000569.1&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;CDM:cntg-e6a45a6a-24ed-474f-85…</td><td>&quot;JAIBEQ010000898.1&quot;</td></tr><tr><td>&quot;CDM:cntg-c0a8dd32-d692-440c-93…</td><td>&quot;JAIBEQ010000903.1&quot;</td></tr><tr><td>&quot;CDM:cntg-184350e7-9c4c-470d-a6…</td><td>&quot;JAIBEQ010000916.1&quot;</td></tr><tr><td>&quot;CDM:cntg-ffca6144-0569-49a6-b2…</td><td>&quot;JAIBEQ010000175.1&quot;</td></tr><tr><td>&quot;CDM:cntg-be1d1b6f-c308-43ce-a0…</td><td>&quot;JAIBEQ010000183.1&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (1_064, 2)\n",
       "┌─────────────────────────────────┬───────────────────┐\n",
       "│ entity_id                       ┆ name              │\n",
       "│ ---                             ┆ ---               │\n",
       "│ str                             ┆ str               │\n",
       "╞═════════════════════════════════╪═══════════════════╡\n",
       "│ CDM:cntg-0652c309-15ab-47fc-9c… ┆ JAIBEQ010000430.1 │\n",
       "│ CDM:cntg-77920f40-360f-4b30-b7… ┆ JAIBEQ010000456.1 │\n",
       "│ CDM:cntg-35e9d20a-1730-471b-8e… ┆ JAIBEQ010000495.1 │\n",
       "│ CDM:cntg-a142609d-adf6-4587-8c… ┆ JAIBEQ010000514.1 │\n",
       "│ CDM:cntg-3cecc5bf-5c5b-440c-b6… ┆ JAIBEQ010000569.1 │\n",
       "│ …                               ┆ …                 │\n",
       "│ CDM:cntg-e6a45a6a-24ed-474f-85… ┆ JAIBEQ010000898.1 │\n",
       "│ CDM:cntg-c0a8dd32-d692-440c-93… ┆ JAIBEQ010000903.1 │\n",
       "│ CDM:cntg-184350e7-9c4c-470d-a6… ┆ JAIBEQ010000916.1 │\n",
       "│ CDM:cntg-ffca6144-0569-49a6-b2… ┆ JAIBEQ010000175.1 │\n",
       "│ CDM:cntg-be1d1b6f-c308-43ce-a0… ┆ JAIBEQ010000183.1 │\n",
       "└─────────────────────────────────┴───────────────────┘"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_contig_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "c9c83864-56ae-4042-81cd-efffd5cac8e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_names = {row['entity_id']: row['name'] for row in df_names.rows(named=True)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "8b4b006b-e3ae-4478-85ea-2889735f25f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (5_983, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>entity_id</th><th>name</th></tr><tr><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;CDM:feat-7789e6da-eab5-46c7-91…</td><td>&quot;JAIBEQ010000430.1_1&quot;</td></tr><tr><td>&quot;CDM:feat-8d9e7d7e-6bdc-4bb9-81…</td><td>&quot;JAIBEQ010000430.1_2&quot;</td></tr><tr><td>&quot;CDM:feat-7923a96d-9def-47cd-80…</td><td>&quot;JAIBEQ010000430.1_3&quot;</td></tr><tr><td>&quot;CDM:feat-ff7fa704-472b-4575-aa…</td><td>&quot;JAIBEQ010000456.1_1&quot;</td></tr><tr><td>&quot;CDM:feat-ad7617c8-9dbe-4004-94…</td><td>&quot;JAIBEQ010000456.1_2&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;CDM:feat-7e1ce321-3940-4130-85…</td><td>&quot;JAIBEQ010000183.1_6&quot;</td></tr><tr><td>&quot;CDM:feat-3b38b767-d1ec-4222-a9…</td><td>&quot;JAIBEQ010000183.1_7&quot;</td></tr><tr><td>&quot;CDM:feat-9983306d-f095-4b65-b1…</td><td>&quot;JAIBEQ010000183.1_8&quot;</td></tr><tr><td>&quot;CDM:feat-9458dd09-5361-4e91-9f…</td><td>&quot;JAIBEQ010000183.1_9&quot;</td></tr><tr><td>&quot;CDM:feat-9dce0702-f965-4600-86…</td><td>&quot;JAIBEQ010000183.1_10&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (5_983, 2)\n",
       "┌─────────────────────────────────┬──────────────────────┐\n",
       "│ entity_id                       ┆ name                 │\n",
       "│ ---                             ┆ ---                  │\n",
       "│ str                             ┆ str                  │\n",
       "╞═════════════════════════════════╪══════════════════════╡\n",
       "│ CDM:feat-7789e6da-eab5-46c7-91… ┆ JAIBEQ010000430.1_1  │\n",
       "│ CDM:feat-8d9e7d7e-6bdc-4bb9-81… ┆ JAIBEQ010000430.1_2  │\n",
       "│ CDM:feat-7923a96d-9def-47cd-80… ┆ JAIBEQ010000430.1_3  │\n",
       "│ CDM:feat-ff7fa704-472b-4575-aa… ┆ JAIBEQ010000456.1_1  │\n",
       "│ CDM:feat-ad7617c8-9dbe-4004-94… ┆ JAIBEQ010000456.1_2  │\n",
       "│ …                               ┆ …                    │\n",
       "│ CDM:feat-7e1ce321-3940-4130-85… ┆ JAIBEQ010000183.1_6  │\n",
       "│ CDM:feat-3b38b767-d1ec-4222-a9… ┆ JAIBEQ010000183.1_7  │\n",
       "│ CDM:feat-9983306d-f095-4b65-b1… ┆ JAIBEQ010000183.1_8  │\n",
       "│ CDM:feat-9458dd09-5361-4e91-9f… ┆ JAIBEQ010000183.1_9  │\n",
       "│ CDM:feat-9dce0702-f965-4600-86… ┆ JAIBEQ010000183.1_10 │\n",
       "└─────────────────────────────────┴──────────────────────┘"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "eb6f8211-a0ac-4e23-8cf9-04b5050d0fd3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CDM:prot-de23b6c3-c0e3-4df9-95e0-0c6e6bd10e6e'"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d_feature_to_protein[cdm_feature_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "94b0c3ce-fc49-4884-a6ef-4675609d3e23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'protein_id': 'CDM:prot-de23b6c3-c0e3-4df9-95e0-0c6e6bd10e6e',\n",
       " 'hash': '1476cb0d2f1f6946951c4bf697ae4aaa7a09cc669ac3350465a5baa45eacc778',\n",
       " 'description': None,\n",
       " 'evidence_for_existence': None,\n",
       " 'length': 714,\n",
       " 'sequence': 'TWRASIIPLLAVPVSVVGTFAVLHLLGFSINALSLFGLVLAIGIVVDDAIVVVENVERNIEAGLTPREATYRAMREVSGPIIAIALVLVAVFVPLAFISGLTGQFYKQFAVTIAISTVISAINSLTLSPALSALLLKGHDEPKDALTRGMDKVFGGLFRGFNRLFHRGSEAYSGGVKRVIGRKALMFVIYLALVGATMGLFKIVPGGFVPAQDKQYLIGFAQLPDGATLDRTENVIRRMGEIMKENPNVEDAIAFPGLSINGFTNSSNSGIVFATLKPFAERTRADQSGGAVAGQLNQSFGSIQDAFIAMFPPPPVAGLGTTGGFKLQIEDRASLGYEAMDNAVKAFMGKAYQTPELAGLFTSWQVNVPQLYANIDRTKARQLGVPVTDIFDTLQIYLGSLYANDFNQFGRTYSVRVQADAAYRARAEDVGLLKVRSATGEMVPLSALMKMEPSFGPERAMRYNGYLAADVNGGPAPGYSSGQAQAAIERIAKETLPQGITFEWTELTYQEILAGNSAVLVFPLAILLVFLVLAAQYESLTLPIAIILIVPMGILAAMTGVWLTKGDNNVFTQIGLIVLVGLSAKNAILIVEFARELEFAGRTPVQAAIEASRLRLRPILMTSLAFVMGVLPLVLSTGAGAEMRSAMGVAVFAGMIGVTAFGLFLTPVFYVLLRKLAGNRPLVEHGAHVAPISHAPGTGGTAHPVLAAPRKEHE'}"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdm_protein['hash']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "b013e58e-cefe-4f4b-a69a-3f06a519b5eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (5_983, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>feature_id</th><th>protein_id</th></tr><tr><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;CDM:feat-7789e6da-eab5-46c7-91…</td><td>&quot;CDM:prot-de23b6c3-c0e3-4df9-95…</td></tr><tr><td>&quot;CDM:feat-8d9e7d7e-6bdc-4bb9-81…</td><td>&quot;CDM:prot-58ae8786-cc2f-4c07-b4…</td></tr><tr><td>&quot;CDM:feat-7923a96d-9def-47cd-80…</td><td>&quot;CDM:prot-fd7bc475-fd90-4ab4-b1…</td></tr><tr><td>&quot;CDM:feat-ff7fa704-472b-4575-aa…</td><td>&quot;CDM:prot-9d28af3f-292d-48b1-8a…</td></tr><tr><td>&quot;CDM:feat-ad7617c8-9dbe-4004-94…</td><td>&quot;CDM:prot-2fdf32aa-ffbf-4e99-93…</td></tr><tr><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;CDM:feat-7e1ce321-3940-4130-85…</td><td>&quot;CDM:prot-52202d52-9bca-4b86-83…</td></tr><tr><td>&quot;CDM:feat-3b38b767-d1ec-4222-a9…</td><td>&quot;CDM:prot-cf6b84ca-5214-4d96-b7…</td></tr><tr><td>&quot;CDM:feat-9983306d-f095-4b65-b1…</td><td>&quot;CDM:prot-5732a142-f44a-4786-be…</td></tr><tr><td>&quot;CDM:feat-9458dd09-5361-4e91-9f…</td><td>&quot;CDM:prot-154d7421-1dea-44ca-9e…</td></tr><tr><td>&quot;CDM:feat-9dce0702-f965-4600-86…</td><td>&quot;CDM:prot-48929095-c19a-456c-a6…</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (5_983, 2)\n",
       "┌─────────────────────────────────┬─────────────────────────────────┐\n",
       "│ feature_id                      ┆ protein_id                      │\n",
       "│ ---                             ┆ ---                             │\n",
       "│ str                             ┆ str                             │\n",
       "╞═════════════════════════════════╪═════════════════════════════════╡\n",
       "│ CDM:feat-7789e6da-eab5-46c7-91… ┆ CDM:prot-de23b6c3-c0e3-4df9-95… │\n",
       "│ CDM:feat-8d9e7d7e-6bdc-4bb9-81… ┆ CDM:prot-58ae8786-cc2f-4c07-b4… │\n",
       "│ CDM:feat-7923a96d-9def-47cd-80… ┆ CDM:prot-fd7bc475-fd90-4ab4-b1… │\n",
       "│ CDM:feat-ff7fa704-472b-4575-aa… ┆ CDM:prot-9d28af3f-292d-48b1-8a… │\n",
       "│ CDM:feat-ad7617c8-9dbe-4004-94… ┆ CDM:prot-2fdf32aa-ffbf-4e99-93… │\n",
       "│ …                               ┆ …                               │\n",
       "│ CDM:feat-7e1ce321-3940-4130-85… ┆ CDM:prot-52202d52-9bca-4b86-83… │\n",
       "│ CDM:feat-3b38b767-d1ec-4222-a9… ┆ CDM:prot-cf6b84ca-5214-4d96-b7… │\n",
       "│ CDM:feat-9983306d-f095-4b65-b1… ┆ CDM:prot-5732a142-f44a-4786-be… │\n",
       "│ CDM:feat-9458dd09-5361-4e91-9f… ┆ CDM:prot-154d7421-1dea-44ca-9e… │\n",
       "│ CDM:feat-9dce0702-f965-4600-86… ┆ CDM:prot-48929095-c19a-456c-a6… │\n",
       "└─────────────────────────────────┴─────────────────────────────────┘"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_feature_to_protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b6355c4f-c71c-43b9-8ea9-544203979e1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = ldf_ccol_x_protein.filter(pl.col(\"contig_collection_id\") == ccol_id).select(pl.col(\"protein_id\")).collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "7f5653c4-96f1-471e-a1e3-21ea8493a4c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cdm_prot_ids = {o[0] for o in df.rows()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "a01e55c8-50eb-465b-928b-157cbb16c1d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_proteins = ldf_protein.filter(pl.col(\"protein_id\").is_in(cdm_prot_ids)).collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "c54251e3-57b7-430a-a058-78e245984ffa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = ldf_ccol_x_feature.filter(pl.col(\"contig_collection_id\") == ccol_id).select(pl.col(\"feature_id\")).collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "57dc0aec-a74a-4c9b-87d9-b2be379b0893",
   "metadata": {},
   "outputs": [],
   "source": [
    "cdm_feat_ids = {o[0] for o in df.rows()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "24bc6d9c-a08c-4e29-bef6-816b95fd7952",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5983"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cdm_feat_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "e47da4fe-28a8-4ee8-9483-2180d429dc6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Schema([('entity_id', String),\n",
       "        ('name', String),\n",
       "        ('description', String),\n",
       "        ('source', String)])"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query_genome.ldf_name.collect_schema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d80c4f8-eaa8-4e6f-8fc7-7c87beb3d18e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "1071dbd9-72f7-4b49-a77d-b3d34ca4b2b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_features = ldf_feature.filter(pl.col(\"feature_id\").is_in(cdm_feat_ids)).select([\"feature_id\", \"start\", \"end\", \"type\"]).collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "80f59630-0a7c-41ff-a35d-9d9f4bb1a995",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (5_983, 4)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>feature_id</th><th>start</th><th>end</th><th>type</th></tr><tr><td>str</td><td>i64</td><td>i64</td><td>str</td></tr></thead><tbody><tr><td>&quot;CDM:feat-7789e6da-eab5-46c7-91…</td><td>1</td><td>2145</td><td>&quot;CDS&quot;</td></tr><tr><td>&quot;CDM:feat-8d9e7d7e-6bdc-4bb9-81…</td><td>2145</td><td>2249</td><td>&quot;CDS&quot;</td></tr><tr><td>&quot;CDM:feat-7923a96d-9def-47cd-80…</td><td>2246</td><td>2701</td><td>&quot;CDS&quot;</td></tr><tr><td>&quot;CDM:feat-ff7fa704-472b-4575-aa…</td><td>3</td><td>278</td><td>&quot;CDS&quot;</td></tr><tr><td>&quot;CDM:feat-ad7617c8-9dbe-4004-94…</td><td>275</td><td>892</td><td>&quot;CDS&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;CDM:feat-7e1ce321-3940-4130-85…</td><td>3725</td><td>4522</td><td>&quot;CDS&quot;</td></tr><tr><td>&quot;CDM:feat-3b38b767-d1ec-4222-a9…</td><td>4649</td><td>5614</td><td>&quot;CDS&quot;</td></tr><tr><td>&quot;CDM:feat-9983306d-f095-4b65-b1…</td><td>5671</td><td>6381</td><td>&quot;CDS&quot;</td></tr><tr><td>&quot;CDM:feat-9458dd09-5361-4e91-9f…</td><td>6378</td><td>7655</td><td>&quot;CDS&quot;</td></tr><tr><td>&quot;CDM:feat-9dce0702-f965-4600-86…</td><td>7652</td><td>8386</td><td>&quot;CDS&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (5_983, 4)\n",
       "┌─────────────────────────────────┬───────┬──────┬──────┐\n",
       "│ feature_id                      ┆ start ┆ end  ┆ type │\n",
       "│ ---                             ┆ ---   ┆ ---  ┆ ---  │\n",
       "│ str                             ┆ i64   ┆ i64  ┆ str  │\n",
       "╞═════════════════════════════════╪═══════╪══════╪══════╡\n",
       "│ CDM:feat-7789e6da-eab5-46c7-91… ┆ 1     ┆ 2145 ┆ CDS  │\n",
       "│ CDM:feat-8d9e7d7e-6bdc-4bb9-81… ┆ 2145  ┆ 2249 ┆ CDS  │\n",
       "│ CDM:feat-7923a96d-9def-47cd-80… ┆ 2246  ┆ 2701 ┆ CDS  │\n",
       "│ CDM:feat-ff7fa704-472b-4575-aa… ┆ 3     ┆ 278  ┆ CDS  │\n",
       "│ CDM:feat-ad7617c8-9dbe-4004-94… ┆ 275   ┆ 892  ┆ CDS  │\n",
       "│ …                               ┆ …     ┆ …    ┆ …    │\n",
       "│ CDM:feat-7e1ce321-3940-4130-85… ┆ 3725  ┆ 4522 ┆ CDS  │\n",
       "│ CDM:feat-3b38b767-d1ec-4222-a9… ┆ 4649  ┆ 5614 ┆ CDS  │\n",
       "│ CDM:feat-9983306d-f095-4b65-b1… ┆ 5671  ┆ 6381 ┆ CDS  │\n",
       "│ CDM:feat-9458dd09-5361-4e91-9f… ┆ 6378  ┆ 7655 ┆ CDS  │\n",
       "│ CDM:feat-9dce0702-f965-4600-86… ┆ 7652  ┆ 8386 ┆ CDS  │\n",
       "└─────────────────────────────────┴───────┴──────┴──────┘"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a4432dd-a273-4835-b22a-ac1734e5a55e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (berdl_genomes)",
   "language": "python",
   "name": "berdl_genomes"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
